src/base/bittorrent/customstorage.cpp

```diff
@@ -35,6 +35,191 @@
+{
+}
+
+lt::storage_holder customDiskIOThread::new_torrent(const lt::storage_params &storageParams, const std::shared_ptr<void> &torrent)
+{
+    lt::storage_holder storageHolder = m_nativeDiskIO->new_torrent(storageParams, torrent);
+
+    const QString savePath = Utils::Fs::expandPathAbs(QString::fromStdString(storageParams.path));
+    m_storageData[static_cast<lt::storage_index_t>(storageHolder)] = {
+            savePath
+            , storageParams.mapped_files ? *storageParams.mapped_files : storageParams.files
+            , storageParams.priorities};
+
+    return storageHolder;
+}
+
+void customDiskIOThread::remove_torrent(lt::storage_index_t storage)
+{
+    m_nativeDiskIO->remove_torrent(storage);
+}
+
+void customDiskIOThread::async_read(lt::storage_index_t storage, const lt::peer_request &peerRequest
+                                    , std::function<void (lt::disk_buffer_holder, const lt::storage_error &)> handler
+                                    , lt::disk_job_flags_t flags)
+{
+    m_nativeDiskIO->async_read(storage, peerRequest, handler, flags);
+}
+
+bool customDiskIOThread::async_write(lt::storage_index_t storage, const lt::peer_request &peerRequest
+                                     , const char *buf, std::shared_ptr<lt::disk_observer> diskObserver
+                                     , std::function<void (const lt::storage_error &)> handler, lt::disk_job_flags_t flags)
+{
+    return m_nativeDiskIO->async_write(storage, peerRequest, buf, diskObserver, handler, flags);
+}
+
+void customDiskIOThread::async_hash(lt::storage_index_t storage, lt::piece_index_t piece
+                                    , lt::span<lt::sha256_hash> hash, lt::disk_job_flags_t flags
+                                    , std::function<void (lt::piece_index_t, const lt::sha1_hash &, const lt::storage_error &)> handler)
+{
+    m_nativeDiskIO->async_hash(storage, piece, hash, flags, handler);
+}
+
+void customDiskIOThread::async_hash2(lt::storage_index_t storage, lt::piece_index_t piece
+                                     , int offset, lt::disk_job_flags_t flags
+                                     , std::function<void (lt::piece_index_t, const lt::sha256_hash &, const lt::storage_error &)> handler)
+{
+    m_nativeDiskIO->async_hash2(storage, piece, offset, flags, handler);
+}
+
+void customDiskIOThread::async_move_storage(lt::storage_index_t storage, std::string path, lt::move_flags_t flags
+                                            , std::function<void (lt::status_t, const std::string &, const lt::storage_error &)> handler)
+{
+    const QString newSavePath {Utils::Fs::expandPathAbs(QString::fromStdString(path))};
+
+    if (flags == lt::move_flags_t::dont_replace)
+        handleCompleteFiles(storage, newSavePath);
+
+    m_nativeDiskIO->async_move_storage(storage, path, flags
+                                       , [=](lt::status_t status, const std::string &path, const lt::storage_error &error)
```

Comment:
you may want to move the handler into this lambda too. `[h = std::move(handler)] { ... h(status, path, error); }`